lib/types.js

```diff
@@ -715,11 +715,18 @@ Type.prototype._createBranchConstructor = function () {
   if (name === 'null') {
     return null;
   }
-  var attr = ~name.indexOf('.') ? 'this[\'' + name + '\']' : 'this.' + name;
-  var body = 'return function Branch$(val) { ' + attr + ' = val; };';
-  var Branch = (new Function(body))();
+  function ConstructorFunction() {
+    return function Branch$(val) {
+      if (~name.indexOf('.')) {
+        this[`${name}`] = val;
```

Comment:
@mtth It is easier to remove the usage of template literals and other es6 features but one of the core parts of the refactor is the use of spread operator to pass in the arguments without any function signature change. If not for it we will be forced to change the signature which I was not sure would be better.

For instance, in the existing implementation we do something like,

```
 body += '  return new ' + uname + '(' + innerArgs.join() + ');\n};';
```
this will eventually resolve to something like,
```
return new Person(arg1, arg2, arg3);
```
However when we want to flatten it out into a function we would not able to refactor it as,
```
innerArgs = [arg1, arg2, arg3];
return new self._constructor(innerArgs)
```
we need the spread operator to maintain the same implementation like,
```
innerArgs = [arg1, arg2, arg3]
return new self._constructor(...innerArgs);
```

Let me know if it makes sense. I am open to other ideas as well if this something that can be solved with js that is compatible with Node 0.11

Reply:
Would you be open to the idea of introducing a build step which will compile and build files that are compatible for Node 0.11? Just throwing it out to see if thats a possibility that we are open to.

Reply:
You can emulate the spread operator by `apply`ing on `bind`:

```javascript
function Foo(a, b) { this.a = a; this.b = b; } // Example constructor.
var args = [undefined, 1, 2]; // The dynamic array of arguments to apply.
var BoundFoo = Foo.bind.apply(Foo, args); // Bind the array to the constructor.
var foo = new BoundFoo(); // Foo { a: 1, b: 2 }
```

Introducing a build step sounds acceptable if it doesn't introduce too much complexity. It should be done separately from this PR though.